Reorganize the `run-make-support` library

jieyouxu opened the pull request:
The `run_make_support` library has a kitchen sink `lib.rs` that make discovery/learning very difficult. Let's try to improve that by breaking up `lib.rs` into smaller more organized modules. This is a precursor to improving the documentation and learnability of the `run_make_support` library.

### Changes

- Breakup `lib.rs` into smaller modules according to functionality
- Rename `recursive_diff` -> `assert_dirs_are_equal`
- Rename one of the `read_dir` with callback interface as `read_dir_entries`
- Coalesced fs-related stuff onto a `fs` module, re-exported to tests as `rfs`
- Minor doc improvements / fixes in a few places (I have a follow-up documentation PR planned)

This PR is best reviewed commit-by-commit.

r? @Kobzol (or Mark, or T-compiler or T-bootstrap)

try-job: x86_64-msvc
try-job: aarch64-apple
try-job: test-various
try-job: armhf-gnu
try-job: dist-x86_64-linux

jieyouxu committed:
run_make_support: move `impl_common_helpers` into own `macros` module

jieyouxu committed:
run_make_support: move target checks into `targets` module

jieyouxu committed:
run_make_support: move `env_var{,_os}` into `env_checked` module

jieyouxu committed:
run_make_support: move external deps to `external_deps` module

jieyouxu committed:
run_make_support: cleanup and document some lib.rs reexports

jieyouxu committed:
run_make_support: move `ar` into own module

jieyouxu committed:
run_make_support: move artifact name helpers into `artifact_names` module

jieyouxu committed:
run_make_support: move path-related helpers into own module

jieyouxu committed:
run_make_support: move fs helpers to own module

jieyouxu committed:
run_make_support: move `run_in_tmpdir` and `test_while_readonly` to `scoped_run` module

jieyouxu committed:
run_make_support: move assertions and helpers into own module

jieyouxu committed:
run_make_support: rename `recursive_diff` to `assert_recursive_eq`

jieyouxu committed:
run_make_support: move `assert_recursive_eq` into `assertion_helpers`

jieyouxu committed:
run_make_support: move `handle_failed_output` into `util` module

jieyouxu committed:
run_make_support: make `set_host_rpath` private and move into `util`

jieyouxu committed:
tests: update rustdoc test for renamed `assert_recursive_eq`

jieyouxu commented:
I have ran `./x doc run-make-support` locally, but let's also run some try jobs to be sure.
@bors try

bors commented:
:hourglass: Trying commit 4769fc9669effba0a31cab92faa37c6025ea193c with merge e8f1965e243dd906e4b2561e47b263a887f5712f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4769fc9669effba0a31cab92faa37c6025ea193c","merge_sha":"e8f1965e243dd906e4b2561e47b263a887f5712f"} -->

jieyouxu commented:
@rustbot author (while waiting for try job to come back)

jieyouxu commented:
This will conflict (and is prone to conflicts) with other rmake.rs PRs.
@bors rollup=never p=1

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9939316694/job/27459241289)
Build commit: e8f1965e243dd906e4b2561e47b263a887f5712f (`e8f1965e243dd906e4b2561e47b263a887f5712f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9939316694/job/27459241289"},"merge_sha":"e8f1965e243dd906e4b2561e47b263a887f5712f"} -->

jieyouxu commented:
Try builds passed.
@rustbot ready

fmease commented on src/tools/run-make-support/src/artifact_names.rs:
```diff
@@ -0,0 +1,81 @@
+
+/// Construct the binary (executable) name based on the target.
+#[must_use]
+pub fn bin_name(name: &str) -> String {
```
(preexisting) this could use `std::env::consts::EXE_EXTENSION`

jieyouxu replied:
Thanks for pointing that out 👍, I'll address these in a follow-up PR. I want to keep this PR mostly just moving code around.

fmease commented on src/tools/run-make-support/src/artifact_names.rs:
```diff
@@ -0,0 +1,81 @@
+
+/// Construct the dynamic library extension based on the target.
+#[must_use]
+pub fn dynamic_lib_extension() -> &'static str {
```
(preexisting) this could be replaced with `std::env::consts::DLL_EXTENSION`

Kobzol commented on src/tools/run-make-support/src/macros.rs:
````diff
@@ -0,0 +1,113 @@
+/// ```ignore (illustrative)
+/// struct CommandWrapper { cmd: Command } // <- required `cmd` field
+///
+/// crate::macors::impl_common_helpers!(CommandWrapper);
````
`macors` -> `macros`

jieyouxu replied:
Fixed

Kobzol commented on src/tools/run-make-support/src/env_checked.rs:
```diff
@@ -0,0 +1,20 @@
+use std::ffi::OsString;
```
Why the name `checked`? We do check/panic/assert pretty much everything in `run-make-support`, but the other modules are not named like this.

jieyouxu replied:
Naming is hard. Renamed to `env`.

Kobzol commented on src/tools/run-make-support/src/lib.rs:
```diff
@@ -40,7 +41,13 @@ pub use python::python_command;
 pub use rustdoc::{bare_rustdoc, rustdoc, Rustdoc};
 
-/// [`diff`] is implemented in terms of the [similar] library.
+/// [`ar`][mod@ar] currently uses the [ar][rust-ar] rust library, but that is subject to changes, we
```
`mod@ar` is some `rustdoc` syntax? :eyes: That's the first time I'm seeing this :laughing:

jieyouxu replied:
Yes, it's required (at the time) to distinguish `ar` the module (type namespace) versus `ar` the function (value namespace).

Kobzol commented on src/tools/run-make-support/src/path_helpers.rs:
```diff
@@ -0,0 +1,66 @@
+/// > -- *from [cygpath documentation](https://cygwin.com/cygwin-ug-net/cygpath.html)*.
+#[track_caller]
+#[must_use]
+pub fn cygpath_windows<P: AsRef<Path>>(path: P) -> String {
```
I would either move this to `external_deps`, or rename the function. First I thought that this returns a path to the `cygpath` binary.

Maybe something like `get_windows_path`? `cygpath` is an implementation detail anyway, right?

jieyouxu replied:
Moved to `external_deps`, made library-private, and renamed to `get_windows_path`.

Kobzol commented on src/tools/run-make-support/src/fs_helpers.rs:
```diff
@@ -0,0 +1,71 @@
+use std::io;
```
I don't think we need two modules, it's a bit confusing to decide between `fs_helpers` and `fs_wrapper`. I'd create just a single module, and we can also just name it `fs`, to make tests shorter, seeing `fs_wrapper` everywhere is a bit distracting.

Kobzol commented on src/tools/run-make-support/src/assertion_helpers.rs:
```diff
@@ -0,0 +1,142 @@
+/// Browse the directory `path` non-recursively and return all files which respect the parameters
+/// outlined by `closure`.
+#[track_caller]
+pub fn shallow_find_files<P: AsRef<Path>, F: Fn(&PathBuf) -> bool>(
```
I would put this under a `path` helper, or a `fs` helper.

jieyouxu replied:
Moved to `path_heleprs`.

Kobzol commented on src/tools/run-make-support/src/assertion_helpers.rs:
```diff
@@ -0,0 +1,142 @@
+}
+
+/// Returns true if the filename at `path` starts with `prefix`.
+pub fn has_prefix<P: AsRef<Path>>(path: P, prefix: &str) -> bool {
```
Same here, I would put these under path or fs.

jieyouxu replied:
Moved to `path_heleprs`.

Kobzol commented on src/tools/run-make-support/src/assertion_helpers.rs:
```diff
@@ -0,0 +1,142 @@
+
+/// Gathers all files in the current working directory that have the extension `ext`, and counts
+/// the number of lines within that contain a match with the regex pattern `re`.
+pub fn count_regex_matches_in_files_with_extension(re: &regex::Regex, ext: &str) -> usize {
```
I would put these under some string module. It seems like a better split to me, `assertion_helpers` is quite general, and you can use some of these functions also for other things than assertions.

jieyouxu replied:
Moved to new `string` module.

Kobzol commented on src/tools/run-make-support/src/lib.rs:
```diff
@@ -117,13 +117,13 @@ pub fn set_host_rpath(cmd: &mut Command) {
-/// Check that all files in `dir1` exist and have the same content in `dir2`. Panic otherwise.
-pub fn recursive_diff(dir1: impl AsRef<Path>, dir2: impl AsRef<Path>) {
+/// Assert that all files in `dir1` exist and have the same content in `dir2`
+pub fn assert_recursive_eq(dir1: impl AsRef<Path>, dir2: impl AsRef<Path>) {
```
This name does not make it immediately clear that we're talking about directories. Maybe `assert_dirs_are_equal`?

jieyouxu replied:
Renamed to `assert_dirs_are_equal` and updated 1 test.

Kobzol commented on src/tools/run-make-support/src/assertion_helpers.rs:
```diff
@@ -140,3 +141,28 @@ pub fn assert_not_contains<H: AsRef<str>, N: AsRef<str>>(haystack: H, needle: N)
     }
 }
+
+/// Assert that all files in `dir1` exist and have the same content in `dir2`
```
This *is* an actual assertion helper, since it does assert inside :)

Kobzol reviewed (commented):
Very nice cleanup! Left a few comments, pretty much all nits.

Btw https://github.com/tummychow/git-absorb might be useful if you want to make changes to so many commits.

jieyouxu committed:
run_make_support: move `build_native_static_lib` under `external_deps`

And fix some `ar`-related rebase mishaps.

jieyouxu committed:
run_make_support: rename `env_checked` -> `env`

jieyouxu committed:
run_make_support: rename `cygpath_windows` to `get_windows_path` and move under `external_deps` as private

jieyouxu committed:
run_make_support: coalesce fs helpers into single `fs` module

There were *two* `read_dir` helpers, one being a simple
`std::fs::read_dir` wrapper, the other has a different callback-based
signature. We also rename the callback-based `read_dir` as
`read_dir_entries`.

Also don't top-level re-export most `fs::*` helpers.

jieyouxu committed:
tests: update for renamed `fs` module in run_make_support

jieyouxu committed:
run_make_support: move some helpers from assertion to path

jieyouxu committed:
run_make_support: moved some helpers into `string` module

jieyouxu committed:
run_make_support: rename `assert_recursive_eq` to `assert_dirs_are_equal`

jieyouxu committed:
tests: update rustdoc test for renamed `assert_dirs_are_equal`

jieyouxu commented:
### Changes since last review

- Fixed typos
- Renamed `env_checked` module to `env`
- Moved `cygpath_windows` to `external_deps` and made library-internal
- Coalesced `fs_wrapper` and `fs_helpers` into single `fs` module
    - Updated *several* tests accordingly
- Moved some helpers out of `assertion_helpers` into `path_helpers` or `fs`
- Moved `count_regex_matches_in_files_with_extension` and `invalid_utf8{_,_not_}contains` to new `string` module
- Renamed `assert_recursive_eq` to `assert_dirs_are_equal`
    - Updated 1 rustdoc test that used it

jieyouxu commented:
Trying again just to be safe.
@rustbot author
@bors try

bors commented:
:hourglass: Trying commit 474488cd4e3c7e4693dd08ba62dc87af5be63655 with merge 21ca636bd02fd30e554b49d63028c620f956291a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"474488cd4e3c7e4693dd08ba62dc87af5be63655","merge_sha":"21ca636bd02fd30e554b49d63028c620f956291a"} -->

Kobzol commented:
Thanks a lot! One remaining nit: I think it's nice that the tests use the `rfs` prefix to make it clear that they don't use the `std::fs` module. But it seems a bit annoying and easy to forget to use `use run_make_support::{fs as rfs}` in each runmake test. What about just doing `mod fs; pub use fs as rfs;` in `run-make-support`, so that the module is named `fs` internally (to be consistent with other modules), but `rfs` in `run-make-support` (to distinguish it from `std::fs`, but also without requiring each test to use `fs as rfs`)?

jieyouxu commented:
> Thanks a lot! One remaining nit: I think it's nice that the tests use the `rfs` prefix to make it clear that they don't use the `std::fs` module. But it seems a bit annoying and easy to forget to use `use run_make_support::{fs as rfs}` in each runmake test. What about just doing `mod fs; pub use fs as rfs;` in `run-make-support`, so that the module is named `fs` internally in `run-make-support` (to be consistent with other modules), but `rfs` in run-make tests (to distinguish it from `std::fs`, but also without requiring each test to use `fs as rfs`)?

Sounds reasonable, I'll do that and update the tests.

Kobzol commented:
Thanks! After that change, and with tests green, feel free to r=me.

jieyouxu committed:
run_make_support: use `fs` internally, but `rfs` to tests

jieyouxu committed:
tests: update for `rfs` rename

jieyouxu commented:
### Changes since last review

- Renamed to `fs` locally but re-exported as `rfs` to tests
    - Updated tests accordingly

jieyouxu commented:
Going to re-run try jobs (I think force push cancels them?)
@bors try

Kobzol commented:
> I think force push cancels them?

It doesn't, we currently allow parallel try builds on PRs.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9975306432/job/27570804457)
Build commit: 1464be56c803c09ec9575ff8db7d1a83020da727 (`1464be56c803c09ec9575ff8db7d1a83020da727`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9975306432/job/27570804457"},"merge_sha":"1464be56c803c09ec9575ff8db7d1a83020da727"} -->

jieyouxu commented:
@bors r=@Kobzol

bors commented:
:pushpin: Commit d69cc1ccbf9d25225add138a4b0f6ee424d0d59a has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol d69cc1ccbf9d25225add138a4b0f6ee424d0d59a -->
<!-- homu: {"type":"Approved","sha":"d69cc1ccbf9d25225add138a4b0f6ee424d0d59a","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
(RIP all the open run-make PRs)

bors commented:
:hourglass: Testing commit d69cc1ccbf9d25225add138a4b0f6ee424d0d59a with merge f00f85091904530ee8f5b29251c53571e0062a76...
<!-- homu: {"type":"BuildStarted","head_sha":"d69cc1ccbf9d25225add138a4b0f6ee424d0d59a","merge_sha":"f00f85091904530ee8f5b29251c53571e0062a76"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9977432089/job/27578372302)
Approved by: Kobzol
Pushing f00f85091904530ee8f5b29251c53571e0062a76 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9977432089/job/27578372302"},"merge_sha":"f00f85091904530ee8f5b29251c53571e0062a76"} -->

rust-timer commented:
Finished benchmarking commit (f00f85091904530ee8f5b29251c53571e0062a76): [comparison URL](https://perf.rust-lang.org/compare.html?start=08cdc2fa1a9fd5ba466838f69cfaf062a3a64ad1&end=f00f85091904530ee8f5b29251c53571e0062a76&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=08cdc2fa1a9fd5ba466838f69cfaf062a3a64ad1&end=f00f85091904530ee8f5b29251c53571e0062a76&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=08cdc2fa1a9fd5ba466838f69cfaf062a3a64ad1&end=f00f85091904530ee8f5b29251c53571e0062a76&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=08cdc2fa1a9fd5ba466838f69cfaf062a3a64ad1&end=f00f85091904530ee8f5b29251c53571e0062a76&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=08cdc2fa1a9fd5ba466838f69cfaf062a3a64ad1&end=f00f85091904530ee8f5b29251c53571e0062a76&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.855s -> 768.932s (-0.12%)
**Artifact size**: 328.79 MiB -> 328.74 MiB (-0.01%)
